NonSucking.Framework.Serialization/NoosonGenerator.cs

```diff
@@ -267,327 +197,494 @@ private static string TypeNameToSummaryName(string typeName)
+
+            generatedType.Methods.Add(new GeneratedMethod(
+                null,
+                Consts.Serialize,
+                parameters,
+                modifiers,
+                generateGeneric
+                    ? new[]
+                    {
+                        new TypeParameter(Consts.GenericParameterWriterName,
+                            xmlDocumentation: "The type of the instance to serialize.")
+                    }
+                    : Array.Empty<TypeParameter>(), generateConstraint
+                    ? new[]
+                    {
+                        new TypeParameterConstraintClause(Consts.GenericParameterWriterName,
+                            new TypeParameterConstraint("NonSucking.Framework.Serialization.IBinaryWriter"))
+                    }
+                    : Array.Empty<TypeParameterConstraintClause>(),
+                body,
+                additionalParameter is null
+                    ? "Serializes this instance."
+                    : $"Serializes the given <see cref=\"{member.TypeSymbol.ToSummaryName()}\"/> instance."));
         }
 
-        internal static BaseMethodDeclarationSyntax GenerateDeserializeMethod(ITypeSymbol typeSymbol, NoosonGeneratorContext context)
+        internal static void GenerateDeserializeMethod(GeneratedType generatedType, ITypeSymbol typeSymbol,
+            NoosonGeneratorContext context)
         {
-            const string genericParameterName = "TNonSuckingReader";
-            
             var generateGeneric = context.ReaderTypeName is null;
-            var typeName = context.ReaderTypeName ?? genericParameterName;
+            var typeName = context.ReaderTypeName ?? Consts.GenericParameterReaderName;
             var member = new MemberInfo(typeSymbol, typeSymbol, ReturnValueBaseName);
-            var parameter = SyntaxFactory.ParseParameterList($"{typeName} {context.ReaderWriterName}");
+            var parameter = new List<GeneratedMethodParameter>()
+                {new GeneratedMethodParameter(typeName, context.ReaderWriterName, new(), $"The <see cref=\"{context.ReaderTypeName}\"/> to deserialize from.")};
+            var methodType = context.MethodType;
+            if (methodType == MethodType.DeserializeIntoInstance)
+            {
+                var parameterInstance = new GeneratedMethodParameter(typeSymbol.ToDisplayString(), Consts.InstanceParameterName, new(), "The instance to deserialize into.");
+                if (typeSymbol.IsValueType)
+                    parameterInstance.Modifier.Add(ParameterModifiers.Ref);
+                parameter.Insert(0, parameterInstance);
+            }
+
             var body
-                = CreateBlock(member, context, MethodType.Deserialize);
-        
-            var modifiers = new List<Modifiers> { Modifiers.Public };
-        
+                = CreateBlock(member, context, methodType);
+
+            var modifiers = new List<Modifiers> { Modifiers.Public, Modifiers.Static };
```

Comment:
Don't we want the context modifiers here instead of hardcoded public(as a copy)?